convex/game/inputs.ts

```diff
@@ -1,425 +1,240 @@
+import * as conversations from './conversations';
 import { point } from '../util/types';
 import { AiTown } from './aiTown';
-import { assertNever } from '../util/assertNever';
-import { Descriptions, characters } from '../../data/characters';
-import { blocked, movePlayer } from './movement';
-import { acceptInvite, leaveConversation, rejectInvite } from './conversationMembers';
-import { startConversation, stopConversation } from './conversations';
-import { activity } from './players';
 
-export const inputs = {
-  // Join, creating a new player...
-  join: {
-    args: v.object({
-      name: v.string(),
-      character: v.string(),
-      description: v.string(),
-      tokenIdentifier: v.optional(v.string()),
-    }),
-    returnValue: v.id('players'),
+export function inputHandler<ArgsValidator extends PropertyValidators, Return extends any>(def: {
+  args: ArgsValidator;
+  handler: (game: AiTown, now: number, args: ObjectType<ArgsValidator>) => Promise<Return>;
+}) {
+  return def;
+}
+// Join, creating a new player...
+
+export const join = inputHandler({
+  args: {
+    name: v.string(),
+    character: v.string(),
+    description: v.string(),
+    tokenIdentifier: v.optional(v.string()),
   },
-  // ...or leave, disabling the specified player.
-  leave: {
-    args: v.object({
-      playerId: v.id('players'),
-    }),
-    returnValue: v.null(),
+  handler: async (game, now, args) => {
+    const { name, character, description, tokenIdentifier } = args;
+    const players = game.players.allDocuments();
+    let position;
+    for (let attempt = 0; attempt < 10; attempt++) {
+      const candidate = {
+        x: Math.floor(Math.random() * game.map.width),
+        y: Math.floor(Math.random() * game.map.height),
+      };
+      if (blocked(game, now, candidate)) {
+        continue;
+      }
+      position = candidate;
+      break;
+    }
+    if (!position) {
+      throw new Error(`Failed to find a free position!`);
+    }
+    const facingOptions = [
+      { dx: 1, dy: 0 },
+      { dx: -1, dy: 0 },
+      { dx: 0, dy: 1 },
+      { dx: 0, dy: -1 },
+    ];
+    const facing = facingOptions[Math.floor(Math.random() * facingOptions.length)];
+    if (!characters.find((c) => c.name === character)) {
+      throw new Error(`Invalid character: ${character}`);
+    }
+    const locationId = await game.locations.insert(now, {
+      x: position.x,
+      y: position.y,
+      dx: facing.dx,
+      dy: facing.dy,
+      velocity: 0,
+    });
+    const playerId = await game.players.insert({
+      worldId: game.world._id,
+      name,
+      description,
+      active: true,
+      human: tokenIdentifier,
+      character,
+      locationId,
+    });
+    return playerId;
   },
-
-  // Move the player to a specified location.
-  moveTo: {
-    args: v.object({
-      playerId: v.id('players'),
-      destination: v.union(point, v.null()),
-    }),
-    returnValue: v.null(),
+} as const);
```

Comment:
ah interesting, why do we need the `as const` here but not in the agent handlers?

Reply:
good catch- this was leftover from trying to get codegen to associate the name and args. It's currently working without this, so I'll remove it